articles/execute_process_refactor.md

```diff
@@ -0,0 +1,596 @@
+
+#### Methods
+
+No methods would be defined or overridden.
+
+#### Events
+
+No events would be handled or emitted.
+
+### launch_ros.traits.NodeTrait
+This abstract class would not extend any others, and would provide a common interface to allow definition of traits which apply to `launch_ros.descriptions.Node` objects.
+
+#### Constructor
+
+This constructor would take no parameters, and should not be used directly.
+
+#### Properties
+
+This object would expose no properties.
+
+#### Methods
+
+To allow a trait to perform its actions, the `prepare()` method will be called prior to execution.
+
+| Name    | Description                                                  |
+| ------- | ------------------------------------------------------------ |
+| prepare | Abstract method which takes a given Node, LaunchContext, and Action, and performs actions necessary to prepare the node for execution in that context.|
+
+#### Events
+
+No events would be handled or emitted.
+
+### launch_ros.traits.HasLifecycle
+
+This class would extend the `launch_ros.traits.NodeTrait` class, and provide the additional functionality required for managing lifecycle node events that are currently defined in `launch_ros.actions.LifecycleNode`. 
+
+#### Constructor
+
+The constructor for this trait would take no parameters, as it serves only to assert that the ROS2 Lifecycle events are implemented by the `launch_ros.descriptions.Node`, and appropriate events should be managed.
+
+#### Properties
+
+This object would expose no properties.
+
+#### Methods
+
+| Name    | Description                                                  |
+| ------- | ------------------------------------------------------------ |
+| prepare | Overridden. Adds subscription event handlers to listen for ROS node lifecycle state changes; this handler will trigger appropriate events from this class when transitions are detected. |
+
+#### Events
+
+##### Handled Events
+
+| Event Type                            | Description                                                  |
+| ------------------------------------- | ------------------------------------------------------------ |
+| `launch.events.lifecycle.ChangeState` | This event can be targeted to a single lifecycle node, or more than one, or even all lifecycle nodes, and it requests the targeted nodes to change state, see its documentation for more details |
+
+##### Emitted Events
+
+| Event Type                                | Description                                                  |
+| ----------------------------------------- | ------------------------------------------------------------ |
+| `launch.events.lifecycle.StateTransition` | This event is emitted when a message is published to the "`/<node_name>/transition_event`" topic, indicating the lifecycle node represented by this action changed state |
+
+### launch_ros.traits.ComposesNodes
```

Comment:
ComposesNodes sound weird to me.
It sounds like a ComposableNode container (?)

Reply:
Yes, that's the function here. But in trying to go with the "traits" concept, I was looking for naming what was more along the lines of "a thing this class does". I'm certainly open to changing the name if there's something that sounds better. I was a little worried that going straight with `launch_ros.traits.ComposableNodeContainer` might make it seem like this is the thing itself, rather than describing an aspect of the `launch_ros.descriptions.Node`.

Reply:
I agree with @roger-strain. I can't think of a better name either. Ultimately, there will be a handy `ComposableNodeContainer` action that hides this away, so I don't see a risk of confusion.